### PR TITLE
`OpacityEffect` in the new effects engine

### PR DESCRIPTION
**Pull request by spydon**
# Description

The `OpacityEffect` is a bit tricky to compose since the opacity only is stored in 8 bits, so it can get pretty big numeric discrepancies, so we have to accumulate the rounding errors. 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `flutter format` and the `flutter analyze` does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database]. Indicate, which of these issues are resolved or fixed by this PR.*

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

**Comment by spydon** [comment]
@st-pasha any idea how we could handle rounding errors like the ones we will get in here when the affected storage is only 8 bits, so that the rounding errors can become quite big?

File: examples/lib/stories/effects/opacity_effect_example.dart [hunk 1]
```diff
@@ -43,10 +46,10 @@ class OpacityEffectExample extends FlameGame with TapDetector {
   @override
   void onTap() {
     final opacity = sprite.paint.color.opacity;
-    if (opacity == 1) {
-      sprite.add(OpacityEffect.fadeOut());
-    } else if (opacity == 0) {
-      sprite.add(OpacityEffect.fadeIn());
+    if (opacity >= 0.5) {
+      sprite.add(OpacityEffect.fadeOut(StandardEffectController(duration: 1)));
```
Comment (erickzanardo) [inline examples/lib/stories/effects/opacity_effect_example.dart:50]: This line is a little bit long now, I wonder if we couldn't create shortcuts or something (not for this PR ofc)

**Review by erickzanardo — APPROVED** [review]

**Comment by st-pasha** [comment]
> any idea how we could handle rounding errors like the ones we will get in here when the affected storage is only 8 bits, so that the rounding errors can become quite big?

We could store double-precision version of a variable within the effect, and then send into the component the rounded value. Also, we'll need to keep track of any external updates to the component's value, and apply it to our double-precision internal value.

**Comment by spydon** [comment]
> We could store double-precision version of a variable within the effect, and then send into the component the rounded value. Also, we'll need to keep track of any external updates to the component's value, and apply it to our double-precision internal value.

If we just take the delta of the internal value each tick it should be fine without keeping track of other updates, right?

**Comment by st-pasha** [comment]
> If we just take the delta of the internal value each tick it should be fine without keeping track of other updates, right?

There is always a possibility that another effects updates alpha, so need to keep track of that.

**Comment by spydon** [comment]
> There is always a possibility that another effects updates alpha, so need to keep track of that.

So you mean the internal double wouldn't be on the effect, but somewhere else? I was thinking that out would be on the effect, and that we only update with the delta from the last internal double. Then they mix their updates, but maybe it wouldn't be any more acccurate than now, since rounding errors would still have to be taken into consideration when converting it to alpha. Maybe the current solution is good enough?

**Comment by st-pasha** [comment]
We can merge it as-is, if it works; and then think about possible improvements later.

**Pull request merged**